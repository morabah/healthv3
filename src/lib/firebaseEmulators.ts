import { getApp } from 'firebase/app';
import { connectAuthEmulator, getAuth } from 'firebase/auth';
import { connectFirestoreEmulator, getFirestore } from 'firebase/firestore';
import { connectFunctionsEmulator, getFunctions } from 'firebase/functions';
import { connectStorageEmulator, getStorage } from 'firebase/storage';

/**
 * Connect to Firebase emulators in development and test environments
 */
export const connectToEmulators = () => {
  try {
    // Only connect to emulators in development or test environment
    if (process.env.NODE_ENV === 'production') {
      return;
    }

    const app = getApp();
    
    // Auth emulator
    const auth = getAuth(app);
    connectAuthEmulator(auth, 'http://localhost:9099', { disableWarnings: true });
    
    // Firestore emulator
    const firestore = getFirestore(app);
    connectFirestoreEmulator(firestore, 'localhost', 8080);
    
    // Functions emulator
    const functions = getFunctions(app);
    connectFunctionsEmulator(functions, 'localhost', 5001);
    
    // Storage emulator
    const storage = getStorage(app);
    connectStorageEmulator(storage, 'localhost', 9199);
    
    console.log('Connected to Firebase emulators');
  } catch (error) {
    console.error('Error connecting to Firebase emulators:', error);
  }
};

/**
 * Helper function to check if emulators are running
 */
export const areEmulatorsRunning = async (): Promise<boolean> => {
  try {
    // Try to connect to the Emulator UI
    const response = await fetch('http://localhost:4000');
    return response.status === 200;
  } catch (error) {
    return false;
  }
};
