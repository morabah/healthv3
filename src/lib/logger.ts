/**
 * Logging utility for the Health Appointment System
 * Provides standardized logging functions with timestamps and log levels
 */

/**
 * Returns current ISO timestamp for consistent log formatting
 */
const getTimestamp = (): string => {
  return new Date().toISOString();
};

/**
 * Type definition for log parameters to ensure type safety
 */
export type LogParams = {
  message: string;
  data?: unknown;
  context?: string;
};

/**
 * Formats log message with timestamp, level, context, and optional data
 */
const formatLogMessage = (
  level: string,
  { message, data, context }: LogParams
): string => {
  const timestamp = getTimestamp();
  const contextStr = context ? `[${context}] ` : '';
  const baseMessage = `[${timestamp}] [${level}] ${contextStr}${message}`;
  
  if (data !== undefined) {
    return `${baseMessage} ${JSON.stringify(data, null, 2)}`;
  }
  
  return baseMessage;
};

/**
 * Logs informational messages
 * @param params Log parameters including message, optional data, and context
 */
export const logInfo = (params: LogParams): void => {
  console.log(formatLogMessage('INFO', params));
};

/**
 * Logs warning messages
 * @param params Log parameters including message, optional data, and context
 */
export const logWarn = (params: LogParams): void => {
  console.warn(formatLogMessage('WARN', params));
};

/**
 * Logs error messages
 * @param params Log parameters including message, optional data, and context
 */
export const logError = (params: LogParams): void => {
  console.error(formatLogMessage('ERROR', params));
};

/**
 * Logs debug messages only when NEXT_PUBLIC_LOG_LEVEL is set to 'debug'
 * @param params Log parameters including message, optional data, and context
 */
export const logDebug = (params: LogParams): void => {
  if (typeof process !== 'undefined' && 
      process.env.NEXT_PUBLIC_LOG_LEVEL === 'debug') {
    console.debug(formatLogMessage('DEBUG', params));
  }
};

/**
 * Default export for convenient importing
 */
export default {
  logInfo,
  logWarn,
  logError,
  logDebug,
};
