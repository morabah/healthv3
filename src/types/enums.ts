/**
 * Enumerations for the Health Appointment System
 * These enums provide type safety for various status and type fields throughout the application
 */

/**
 * User types in the system
 * PATIENT - Regular users who book appointments
 * DOCTOR - Healthcare providers who receive appointments
 * ADMIN - System administrators with elevated privileges
 */
export enum UserType {
  PATIENT = 'PATIENT',
  DOCTOR = 'DOCTOR',
  ADMIN = 'ADMIN'
}

/**
 * Verification status for users (particularly doctors)
 * PENDING - Initial state, awaiting verification
 * VERIFIED - Successfully verified
 * REJECTED - Verification was rejected
 */
export enum VerificationStatus {
  PENDING = 'PENDING',
  VERIFIED = 'VERIFIED',
  REJECTED = 'REJECTED'
}

/**
 * Appointment status tracking
 * PENDING - Initial state when appointment is created
 * CONFIRMED - Appointment has been confirmed by the doctor
 * CANCELLED - Appointment has been cancelled by either party
 * COMPLETED - Appointment has been completed
 */
export enum AppointmentStatus {
  PENDING = 'PENDING',
  CONFIRMED = 'CONFIRMED',
  CANCELLED = 'CANCELLED',
  COMPLETED = 'COMPLETED'
}
