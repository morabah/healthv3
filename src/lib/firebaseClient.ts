/**
 * Firebase Client SDK Initialization
 * This module initializes and exports Firebase client services for use in the frontend
 */

import { initializeApp, getApps, FirebaseApp } from 'firebase/app';
import { getAuth, Auth } from 'firebase/auth';
import { getFirestore, Firestore } from 'firebase/firestore';
import { getStorage, FirebaseStorage } from 'firebase/storage';
import { getFunctions, Functions } from 'firebase/functions';
import { logInfo, logError } from './logger';
import { connectToEmulators } from './firebaseEmulators';

// Firebase configuration from environment variables
const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
  measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
};

// Singleton instances
let firebaseApp: FirebaseApp;
let firebaseAuth: Auth;
let firebaseFirestore: Firestore;
let firebaseStorage: FirebaseStorage;
let firebaseFunctions: Functions;

/**
 * Initialize Firebase if it hasn't been initialized already
 * @returns The Firebase app instance
 */
export const initializeFirebase = (): FirebaseApp => {
  try {
    if (!firebaseApp && getApps().length === 0) {
      logInfo({
        message: 'Initializing Firebase client SDK',
        context: 'firebaseClient',
      });
      
      // Check if required environment variables are set
      if (!process.env.NEXT_PUBLIC_FIREBASE_API_KEY || !process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID) {
        throw new Error('Firebase environment variables are not properly configured');
      }
      
      firebaseApp = initializeApp(firebaseConfig);
      
      // Connect to emulators in development and test environments
      if (process.env.NODE_ENV !== 'production') {
        connectToEmulators();
      }
      
      logInfo({
        message: 'Firebase client SDK initialized successfully',
        context: 'firebaseClient',
        data: { projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID }
      });
    } else if (!firebaseApp) {
      firebaseApp = getApps()[0];
    }
    
    return firebaseApp;
  } catch (error) {
    logError({
      message: 'Failed to initialize Firebase client SDK',
      context: 'firebaseClient',
      data: { error }
    });
    
    throw error;
  }
};

/**
 * Get Firebase Auth instance
 * @returns Firebase Auth instance
 */
export const getFirebaseAuth = (): Auth => {
  if (!firebaseAuth) {
    const app = initializeFirebase();
    firebaseAuth = getAuth(app);
    logInfo({
      message: 'Firebase Auth initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseAuth;
};

/**
 * Get Firebase Firestore instance
 * @returns Firebase Firestore instance
 */
export const getFirebaseFirestore = (): Firestore => {
  if (!firebaseFirestore) {
    const app = initializeFirebase();
    firebaseFirestore = getFirestore(app);
    logInfo({
      message: 'Firebase Firestore initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseFirestore;
};

/**
 * Get Firebase Storage instance
 * @returns Firebase Storage instance
 */
export const getFirebaseStorage = (): FirebaseStorage => {
  if (!firebaseStorage) {
    const app = initializeFirebase();
    firebaseStorage = getStorage(app);
    logInfo({
      message: 'Firebase Storage initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseStorage;
};

/**
 * Get Firebase Functions instance
 * @returns Firebase Functions instance
 */
export const getFirebaseFunctions = (): Functions => {
  if (!firebaseFunctions) {
    const app = initializeFirebase();
    firebaseFunctions = getFunctions(app);
    logInfo({
      message: 'Firebase Functions initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseFunctions;
};

// Initialize Firebase on module load in browser environments
if (typeof window !== 'undefined') {
  try {
    initializeFirebase();
    logInfo({
      message: 'Firebase auto-initialized in browser environment',
      context: 'firebaseClient'
    });
  } catch (error) {
    logError({
      message: 'Error during Firebase auto-initialization',
      context: 'firebaseClient',
      data: { error }
    });
  }
}

// Default export for convenient importing
export default {
  initializeFirebase,
  getFirebaseAuth,
  getFirebaseFirestore,
  getFirebaseStorage,
  getFirebaseFunctions,
};
