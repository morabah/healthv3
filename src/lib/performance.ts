/**
 * Performance tracking utility for the Health Appointment System
 * Provides a simple way to measure and log execution times
 */

import { logInfo } from './logger';

/**
 * Interface for the object returned by trackPerformance
 */
export interface PerformanceTracker {
  /**
   * Stops the performance tracking and logs the duration
   * @param additionalData Optional data to include in the log
   * @returns The duration in milliseconds
   */
  stop: (additionalData?: Record<string, unknown>) => number;
}

/**
 * Tracks the performance of an operation
 * @param label A descriptive label for the operation being tracked
 * @param context Optional context to include in the log (e.g., 'DatabaseService')
 * @returns An object with a stop() method to end tracking and log the duration
 */
export const trackPerformance = (
  label: string,
  context?: string
): PerformanceTracker => {
  // Use performance.now() for high-resolution timing
  const startTime = performance.now();
  
  return {
    stop: (additionalData?: Record<string, unknown>): number => {
      const endTime = performance.now();
      const duration = endTime - startTime;
      
      // Log the performance data using the logger
      logInfo({
        message: `${label} took ${duration.toFixed(2)} ms`,
        context: context || 'PERF',
        data: {
          duration,
          label,
          ...additionalData
        }
      });
      
      return duration;
    }
  };
};

/**
 * Default export for convenient importing
 */
export default {
  trackPerformance
};
