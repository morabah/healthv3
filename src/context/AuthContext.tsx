/**
 * Authentication Context
 * Provides authentication state and user profile data throughout the application
 */

import React, { createContext, useContext, useEffect, useState } from 'react';
import { User } from 'firebase/auth';
import { doc, getDoc } from 'firebase/firestore';
import { useAuthState } from 'react-firebase-hooks/auth';
import { getFirebaseAuth, getFirebaseFirestore } from '../lib/firebaseClient';
import { UserProfile } from '../types/user';
import { logInfo, logError } from '../lib/logger';
import { trackPerformance } from '../lib/performance';

// Context interface defining what data and functions will be available
interface AuthContextType {
  // Firebase authentication state
  user: User | null;
  loading: boolean;
  error: Error | undefined;
  
  // User profile data from Firestore
  userProfile: UserProfile | null;
  profileLoading: boolean;
  
  // Sign out function
  signOut: () => Promise<void>;
}

// Create the context with a default value
const AuthContext = createContext<AuthContextType>({
  user: null,
  loading: true,
  error: undefined,
  userProfile: null,
  profileLoading: true,
  signOut: async () => {},
});

// Custom hook to use the auth context
export const useAuth = () => useContext(AuthContext);

// Provider component that wraps the app and makes auth object available
export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  // Get Firebase auth instance
  const auth = getFirebaseAuth();
  const firestore = getFirebaseFirestore();
  
  // Use the useAuthState hook to track authentication state
  const [user, loading, error] = useAuthState(auth);
  
  // State for user profile data from Firestore
  const [userProfile, setUserProfile] = useState<UserProfile | null>(null);
  const [profileLoading, setProfileLoading] = useState<boolean>(true);
  
  // Function to sign out
  const signOut = async () => {
    try {
      logInfo({
        message: 'User signing out',
        context: 'AuthContext',
        data: { userId: user?.uid }
      });
      
      await auth.signOut();
      
      // Clear user profile state
      setUserProfile(null);
      
      logInfo({
        message: 'User signed out successfully',
        context: 'AuthContext'
      });
    } catch (error) {
      logError({
        message: 'Error signing out',
        context: 'AuthContext',
        data: { error }
      });
      throw error;
    }
  };
  
  // Effect to fetch user profile data when auth state changes
  useEffect(() => {
    // Skip if still loading or no user
    if (loading || !user) {
      if (!loading && !user) {
        // User is definitely signed out
        setProfileLoading(false);
        setUserProfile(null);
        
        logInfo({
          message: 'No authenticated user',
          context: 'AuthContext'
        });
      }
      return;
    }
    
    const fetchUserProfile = async () => {
      // Start performance tracking
      const perfTracker = trackPerformance('fetchUserProfile', 'AuthContext');
      
      try {
        setProfileLoading(true);
        
        logInfo({
          message: 'Fetching user profile',
          context: 'AuthContext',
          data: { userId: user.uid }
        });
        
        // Get user document from Firestore
        const userDocRef = doc(firestore, 'users', user.uid);
        const userDoc = await getDoc(userDocRef);
        
        if (userDoc.exists()) {
          // User profile exists in Firestore
          const userData = userDoc.data() as UserProfile;
          setUserProfile(userData);
          
          logInfo({
            message: 'User profile fetched successfully',
            context: 'AuthContext',
            data: { userId: user.uid, userType: userData.userType }
          });
        } else {
          // User authenticated but no profile in Firestore
          setUserProfile(null);
          
          logInfo({
            message: 'User authenticated but no profile found',
            context: 'AuthContext',
            data: { userId: user.uid }
          });
        }
      } catch (error) {
        logError({
          message: 'Error fetching user profile',
          context: 'AuthContext',
          data: { userId: user.uid, error }
        });
        
        setUserProfile(null);
      } finally {
        setProfileLoading(false);
        perfTracker.stop({ userId: user.uid });
      }
    };
    
    // Log authentication state change
    logInfo({
      message: 'User authenticated',
      context: 'AuthContext',
      data: { 
        userId: user.uid,
        email: user.email,
        emailVerified: user.emailVerified,
        isAnonymous: user.isAnonymous
      }
    });
    
    // Fetch the user profile
    fetchUserProfile();
  }, [user, loading, firestore]);
  
  // Create the value object that will be provided by the context
  const value = {
    user,
    loading,
    error,
    userProfile,
    profileLoading,
    signOut
  };
  
  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthContext;
