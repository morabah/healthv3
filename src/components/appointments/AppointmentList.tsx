/**
 * AppointmentList Component
 * Displays a list of appointments for the authenticated user
 */
import React, { useEffect, useState } from 'react';
import { collection, query, where, orderBy, getDocs } from 'firebase/firestore';
import { getFirebaseFirestore } from '@/lib/firebaseClient';
import { useAuth } from '@/context/AuthContext';
import { AppointmentStatus, UserType } from '@/types/enums';
import { Appointment } from '@/types/appointment';
import { logInfo, logError } from '@/lib/logger';
import { trackPerformance } from '@/lib/performance';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCalendarAlt, faSpinner, faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';

interface AppointmentListProps {
  status?: AppointmentStatus;
  limit?: number;
}

const AppointmentList: React.FC<AppointmentListProps> = ({ 
  status,
  limit = 10
}) => {
  const { user, userProfile, loading: authLoading } = useAuth();
  const [appointments, setAppointments] = useState<Appointment[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);
  
  useEffect(() => {
    // Skip if still loading auth or no user
    if (authLoading || !user || !userProfile) {
      return;
    }
    
    const fetchAppointments = async () => {
      const perfTracker = trackPerformance('fetchAppointments', 'AppointmentList');
      
      try {
        setLoading(true);
        setError(null);
        
        logInfo({
          message: 'Fetching appointments',
          context: 'AppointmentList',
          data: { userId: user.uid, status }
        });
        
        const firestore = getFirebaseFirestore();
        const appointmentsRef = collection(firestore, 'appointments');
        
        // Build query based on user role
        let appointmentsQuery = query(appointmentsRef);
        
        if (userProfile.userType === UserType.PATIENT) {
          appointmentsQuery = query(
            appointmentsRef, 
            where('patientId', '==', user.uid)
          );
        } else if (userProfile.userType === UserType.DOCTOR) {
          appointmentsQuery = query(
            appointmentsRef, 
            where('doctorId', '==', user.uid)
          );
        }
        
        // Add status filter if provided
        if (status) {
          appointmentsQuery = query(
            appointmentsQuery,
            where('status', '==', status)
          );
        }
        
        // Order by appointment date
        appointmentsQuery = query(
          appointmentsQuery,
          orderBy('appointmentDate', 'desc')
        );
        
        const querySnapshot = await getDocs(appointmentsQuery);
        const appointmentsList: Appointment[] = [];
        
        querySnapshot.forEach((doc) => {
          appointmentsList.push({
            id: doc.id,
            ...doc.data()
          } as Appointment);
        });
        
        // Limit the number of appointments if needed
        const limitedAppointments = appointmentsList.slice(0, limit);
        
        setAppointments(limitedAppointments);
        
        logInfo({
          message: 'Appointments fetched successfully',
          context: 'AppointmentList',
          data: { 
            userId: user.uid, 
            count: limitedAppointments.length 
          }
        });
        
        perfTracker.stop({ count: limitedAppointments.length });
      } catch (error: any) {
        logError({
          message: 'Error fetching appointments',
          context: 'AppointmentList',
          data: { error }
        });
        
        setError('Failed to load appointments. Please try again later.');
        perfTracker.stop({ error: true });
      } finally {
        setLoading(false);
      }
    };
    
    fetchAppointments();
  }, [user, userProfile, authLoading, status, limit]);
  
  if (authLoading) {
    return (
      <div className="text-center py-8" data-testid="loading-auth">
        <FontAwesomeIcon icon={faSpinner} spin className="text-blue-500 text-2xl mb-2" />
        <p>Loading user information...</p>
      </div>
    );
  }
  
  if (!user || !userProfile) {
    return (
      <div className="text-center py-8" data-testid="no-auth">
        <p>Please log in to view your appointments</p>
      </div>
    );
  }
  
  if (loading) {
    return (
      <div className="text-center py-8" data-testid="loading-appointments">
        <FontAwesomeIcon icon={faSpinner} spin className="text-blue-500 text-2xl mb-2" />
        <p>Loading appointments...</p>
      </div>
    );
  }
  
  if (error) {
    return (
      <div className="text-center py-8 text-red-500" data-testid="error-message">
        <FontAwesomeIcon icon={faExclamationTriangle} className="text-2xl mb-2" />
        <p>{error}</p>
      </div>
    );
  }
  
  if (appointments.length === 0) {
    return (
      <div className="text-center py-8" data-testid="no-appointments">
        <FontAwesomeIcon icon={faCalendarAlt} className="text-gray-400 text-2xl mb-2" />
        <p>No appointments found</p>
      </div>
    );
  }
  
  return (
    <div className="appointments-list" data-testid="appointments-list">
      <h2 className="text-xl font-semibold mb-4">
        Your Appointments {status && `(${status})`}
      </h2>
      
      <div className="space-y-4">
        {appointments.map((appointment) => (
          <div 
            key={appointment.id} 
            className="appointment-card border rounded-lg p-4 shadow-sm"
            data-testid="appointment-item"
          >
            <div className="flex justify-between items-start">
              <div>
                <h3 className="font-medium">
                  {appointment.reason || 'Appointment'}
                </h3>
                <p className="text-sm text-gray-600">
                  {new Date(appointment.appointmentDate.seconds * 1000).toLocaleDateString()}
                  {' '}at {appointment.startTime} - {appointment.endTime}
                </p>
              </div>
              <div className="appointment-status">
                <span className={`inline-block px-2 py-1 rounded-full text-xs font-medium ${
                  appointment.status === AppointmentStatus.CONFIRMED ? 'bg-green-100 text-green-800' :
                  appointment.status === AppointmentStatus.PENDING ? 'bg-yellow-100 text-yellow-800' :
                  appointment.status === AppointmentStatus.CANCELLED ? 'bg-red-100 text-red-800' :
                  appointment.status === AppointmentStatus.COMPLETED ? 'bg-blue-100 text-blue-800' :
                  'bg-gray-100 text-gray-800'
                }`}>
                  {appointment.status}
                </span>
              </div>
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};

export default AppointmentList;
