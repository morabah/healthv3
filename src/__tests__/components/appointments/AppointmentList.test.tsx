/**
 * AppointmentList Component Tests
 * Tests the appointment list component with mocked Firestore and Auth
 */
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import AppointmentList from '@/components/appointments/AppointmentList';
import { AppointmentStatus, UserType } from '@/types/enums';
import { useAuth } from '@/context/AuthContext';
import { collection, query, where, orderBy, getDocs } from 'firebase/firestore';
import { getFirebaseFirestore } from '@/lib/firebaseClient';
import { logInfo, logError } from '@/lib/logger';
import { trackPerformance } from '@/lib/performance';

// Mock Firebase Firestore
jest.mock('firebase/firestore', () => ({
  collection: jest.fn(),
  query: jest.fn(),
  where: jest.fn(),
  orderBy: jest.fn(),
  getDocs: jest.fn(),
}));

// Mock Firebase client
jest.mock('@/lib/firebaseClient', () => ({
  getFirebaseFirestore: jest.fn(() => ({})),
}));

// Mock Auth context
jest.mock('@/context/AuthContext', () => ({
  useAuth: jest.fn(),
}));

// Mock logger and performance utilities
jest.mock('@/lib/logger', () => ({
  logInfo: jest.fn(),
  logError: jest.fn(),
}));

jest.mock('@/lib/performance', () => ({
  trackPerformance: jest.fn(() => ({
    stop: jest.fn(),
  })),
}));

// Mock appointment data
const mockAppointments = [
  {
    id: 'appt1',
    patientId: 'patient1',
    doctorId: 'doctor1',
    reason: 'Annual checkup',
    appointmentDate: { seconds: 1714435200, nanoseconds: 0 }, // 2024-04-30
    startTime: '10:00',
    endTime: '10:30',
    status: AppointmentStatus.CONFIRMED,
    createdAt: { seconds: 1713830400, nanoseconds: 0 },
    updatedAt: { seconds: 1713830400, nanoseconds: 0 }
  },
  {
    id: 'appt2',
    patientId: 'patient1',
    doctorId: 'doctor2',
    reason: 'Follow-up consultation',
    appointmentDate: { seconds: 1714521600, nanoseconds: 0 }, // 2024-05-01
    startTime: '14:00',
    endTime: '14:30',
    status: AppointmentStatus.PENDING,
    createdAt: { seconds: 1713830400, nanoseconds: 0 },
    updatedAt: { seconds: 1713830400, nanoseconds: 0 }
  },
  {
    id: 'appt3',
    patientId: 'patient1',
    doctorId: 'doctor1',
    reason: 'Vaccination',
    appointmentDate: { seconds: 1714608000, nanoseconds: 0 }, // 2024-05-02
    startTime: '11:00',
    endTime: '11:15',
    status: AppointmentStatus.CANCELLED,
    createdAt: { seconds: 1713830400, nanoseconds: 0 },
    updatedAt: { seconds: 1713830400, nanoseconds: 0 }
  }
];

describe('AppointmentList Component', () => {
  // Mock query builder functions
  const mockCollection = collection as jest.Mock;
  const mockQuery = query as jest.Mock;
  const mockWhere = where as jest.Mock;
  const mockOrderBy = orderBy as jest.Mock;
  const mockGetDocs = getDocs as jest.Mock;
  const mockGetFirestore = getFirebaseFirestore as jest.Mock;
  const mockUseAuth = useAuth as jest.Mock;
  
  beforeEach(() => {
    jest.clearAllMocks();
    
    // Setup mock query chain
    mockCollection.mockReturnValue('appointmentsRef');
    mockQuery.mockReturnValue('appointmentsQuery');
    mockWhere.mockReturnValue('whereQuery');
    mockOrderBy.mockReturnValue('orderByQuery');
    
    // Setup mock Firestore
    mockGetFirestore.mockReturnValue({});
    
    // Setup mock getDocs response
    mockGetDocs.mockResolvedValue({
      forEach: (callback: (doc: any) => void) => {
        mockAppointments.forEach(appointment => {
          callback({
            id: appointment.id,
            data: () => ({ ...appointment })
          });
        });
      }
    });
  });
  
  it('renders loading state when auth is loading', () => {
    mockUseAuth.mockReturnValue({
      user: null,
      userProfile: null,
      loading: true
    });
    
    render(<AppointmentList />);
    
    expect(screen.getByTestId('loading-auth')).toBeInTheDocument();
    expect(screen.getByText('Loading user information...')).toBeInTheDocument();
  });
  
  it('renders message when user is not authenticated', () => {
    mockUseAuth.mockReturnValue({
      user: null,
      userProfile: null,
      loading: false
    });
    
    render(<AppointmentList />);
    
    expect(screen.getByTestId('no-auth')).toBeInTheDocument();
    expect(screen.getByText('Please log in to view your appointments')).toBeInTheDocument();
  });
  
  it('renders appointments for a patient user', async () => {
    // Mock authenticated patient user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    render(<AppointmentList />);
    
    // Should show loading initially
    expect(screen.getByTestId('loading-appointments')).toBeInTheDocument();
    
    // Wait for appointments to load
    await waitFor(() => {
      expect(screen.getByTestId('appointments-list')).toBeInTheDocument();
    });
    
    // Verify Firestore queries
    expect(mockCollection).toHaveBeenCalledWith({}, 'appointments');
    expect(mockWhere).toHaveBeenCalledWith('patientId', '==', 'patient1');
    expect(mockOrderBy).toHaveBeenCalledWith('appointmentDate', 'desc');
    
    // Verify appointments are displayed
    const appointmentItems = screen.getAllByTestId('appointment-item');
    expect(appointmentItems).toHaveLength(3);
    
    // Check specific appointment content
    expect(screen.getByText('Annual checkup')).toBeInTheDocument();
    expect(screen.getByText('Follow-up consultation')).toBeInTheDocument();
    expect(screen.getByText('Vaccination')).toBeInTheDocument();
    
    // Verify logging
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Fetching appointments',
        context: 'AppointmentList'
      })
    );
    
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Appointments fetched successfully',
        context: 'AppointmentList'
      })
    );
  });
  
  it('renders appointments for a doctor user', async () => {
    // Mock authenticated doctor user
    mockUseAuth.mockReturnValue({
      user: { uid: 'doctor1' },
      userProfile: { userType: UserType.DOCTOR },
      loading: false
    });
    
    render(<AppointmentList />);
    
    // Wait for appointments to load
    await waitFor(() => {
      expect(screen.getByTestId('appointments-list')).toBeInTheDocument();
    });
    
    // Verify Firestore queries
    expect(mockCollection).toHaveBeenCalledWith({}, 'appointments');
    expect(mockWhere).toHaveBeenCalledWith('doctorId', '==', 'doctor1');
    expect(mockOrderBy).toHaveBeenCalledWith('appointmentDate', 'desc');
  });
  
  it('filters appointments by status', async () => {
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    // Filter by CONFIRMED status
    render(<AppointmentList status={AppointmentStatus.CONFIRMED} />);
    
    // Wait for appointments to load
    await waitFor(() => {
      expect(screen.getByTestId('appointments-list')).toBeInTheDocument();
    });
    
    // Verify status filter was applied
    expect(mockWhere).toHaveBeenCalledWith('status', '==', AppointmentStatus.CONFIRMED);
    
    // Verify title includes status
    expect(screen.getByText(`Your Appointments (${AppointmentStatus.CONFIRMED})`)).toBeInTheDocument();
  });
  
  it('limits the number of appointments displayed', async () => {
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    // Set limit to 2
    render(<AppointmentList limit={2} />);
    
    // Wait for appointments to load
    await waitFor(() => {
      expect(screen.getByTestId('appointments-list')).toBeInTheDocument();
    });
    
    // Only 2 appointments should be displayed (out of 3 total)
    const appointmentItems = screen.getAllByTestId('appointment-item');
    expect(appointmentItems).toHaveLength(2);
  });
  
  it('handles Firestore query errors', async () => {
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    // Mock Firestore error
    const mockError = new Error('Firestore query failed');
    mockGetDocs.mockRejectedValueOnce(mockError);
    
    render(<AppointmentList />);
    
    // Wait for error message to appear
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toBeInTheDocument();
    });
    
    // Verify error message
    expect(screen.getByText('Failed to load appointments. Please try again later.')).toBeInTheDocument();
    
    // Verify error logging
    expect(logError).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Error fetching appointments',
        context: 'AppointmentList',
        data: expect.objectContaining({ error: mockError })
      })
    );
  });
  
  it('displays message when no appointments are found', async () => {
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    // Mock empty appointments list
    mockGetDocs.mockResolvedValueOnce({
      forEach: () => {} // No appointments
    });
    
    render(<AppointmentList />);
    
    // Wait for no-appointments message to appear
    await waitFor(() => {
      expect(screen.getByTestId('no-appointments')).toBeInTheDocument();
    });
    
    // Verify message
    expect(screen.getByText('No appointments found')).toBeInTheDocument();
  });
  
  it('matches snapshot', async () => {
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'patient1' },
      userProfile: { userType: UserType.PATIENT },
      loading: false
    });
    
    const { container } = render(<AppointmentList />);
    
    // Wait for appointments to load
    await waitFor(() => {
      expect(screen.getByTestId('appointments-list')).toBeInTheDocument();
    });
    
    expect(container).toMatchSnapshot();
  });
});
