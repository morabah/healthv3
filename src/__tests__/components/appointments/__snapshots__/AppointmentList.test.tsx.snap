// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentList Component matches snapshot 1`] = `
<div>
  <div
    class="appointments-list"
    data-testid="appointments-list"
  >
    <h2
      class="text-xl font-semibold mb-4"
    >
      Your Appointments 
    </h2>
    <div
      class="space-y-4"
    >
      <div
        class="appointment-card border rounded-lg p-4 shadow-sm"
        data-testid="appointment-item"
      >
        <div
          class="flex justify-between items-start"
        >
          <div>
            <h3
              class="font-medium"
            >
              Annual checkup
            </h3>
            <p
              class="text-sm text-gray-600"
            >
              4/30/2024
               
              at 
              10:00
               - 
              10:30
            </p>
          </div>
          <div
            class="appointment-status"
          >
            <span
              class="inline-block px-2 py-1 rounded-full text-xs font-medium bg-green-100 text-green-800"
            >
              CONFIRMED
            </span>
          </div>
        </div>
      </div>
      <div
        class="appointment-card border rounded-lg p-4 shadow-sm"
        data-testid="appointment-item"
      >
        <div
          class="flex justify-between items-start"
        >
          <div>
            <h3
              class="font-medium"
            >
              Follow-up consultation
            </h3>
            <p
              class="text-sm text-gray-600"
            >
              5/1/2024
               
              at 
              14:00
               - 
              14:30
            </p>
          </div>
          <div
            class="appointment-status"
          >
            <span
              class="inline-block px-2 py-1 rounded-full text-xs font-medium bg-yellow-100 text-yellow-800"
            >
              PENDING
            </span>
          </div>
        </div>
      </div>
      <div
        class="appointment-card border rounded-lg p-4 shadow-sm"
        data-testid="appointment-item"
      >
        <div
          class="flex justify-between items-start"
        >
          <div>
            <h3
              class="font-medium"
            >
              Vaccination
            </h3>
            <p
              class="text-sm text-gray-600"
            >
              5/2/2024
               
              at 
              11:00
               - 
              11:15
            </p>
          </div>
          <div
            class="appointment-status"
          >
            <span
              class="inline-block px-2 py-1 rounded-full text-xs font-medium bg-red-100 text-red-800"
            >
              CANCELLED
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
